### PR TITLE
[FEATURE][needs-docs] Move layer or group to top of layer panel

### PR DESCRIPTION
**Pull request by MieWinstrup**
A feature that moves the layer(s) or group(s) to the top of the layer panel. This code implements the feature request: https://issues.qgis.org/issues/16936.

## Description
**The reason behind the new feature:**
This functionality is developed to make it easier to move a layer, layers or group(s) with layers to the top of the layer panel. This is especially a useful feature when the layer panel contains a lot of layers. Without this functionality, a user needs to drag the layer with the mouse to the top of the layer panel. This is cumbersome when the layer panel contains a lot of layers. Furthermore, sometimes it’s a little bit difficult to drop the layer at the correct place at the very top of the layer panel. When I switched from ArcMap to QGIS, this was one of the features that I missed in QGIS, especially when I add a layer to my QGIS-project, and then it is placed somewhere in the layer panel (just above the selected layer), quite often I want to have this layer at the very top – with this new feature, I just click on the layer and select ‘Move to Top’.

**Description of the new feature**
This code adds the ‘Move to Top’-functionality to the context menu in the layer panel. This is the menu that appears when right-clicking on a layer/layers or a group/groups in the layer panel. 
The feature works this way:
-	One layer (not in a group): 
	When the user right clicks on one layer (not in a group) and select ‘Move to Top’, the layer is moved to the very top of the layer panel.
-	Multiple layers (none of these are in a group): 
	When the user has selected more than one layer, right clicks on one of them and select ‘Move to Top’, the selected layers are all move to the very top of the layer panel – and with the same order as before.
-	Group(s) with layers: 
	When the user right clicks on a group with layers and select ‘Move to Top’, the group(s) - with all the layers in the group – is/are moved to the top of the layer panel.
-	One layer or more layers in a group: 
	When the user clicks on one layer or more layers within a group, the layer(s) is/are moved to the top of this group. I think this is the most intuitive way to handle this case. If a user wants to move a layer in a group to the top of the layer panel. The workflow would be first to use the ‘Make top level’-function to move a layer out of the group(s) and then use the ‘Move to top’ to move the layer to the top of the layer panel.

To avoid displaying the functionality where it does not make any purpose – i.e. if user clicks on a layer or a group that is already at the top, the code checks if this is the case and decide whether to show the functionality or not.

Currently the functionality is displayed below the ‘Make top level’ in the context menu. I think it make sense to have this feature below or above ‘Make top level’, but maybe all the different functions sometime in the future should be rearranged into something more intuitive order (I think having Properties as the last one would make more sense – but this is of cause not related to this feature request).

**Note**
The current feature ‘Make top level’ is available when a layer is in a group. When a user clicks on ‘Make top level’ the layer is moved out of the group(s). I suggest that the name of this currently feature is changed from ‘Make top level’ to either ‘Move out of group’ or ‘Ungroup’. I think this name change will make it clearer what the feature does and avoid confusion with the new feature ‘Move to Top’. 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by MieWinstrup** [comment]
If #6665 is approved, then I suggest 'Move to Top' should be just above or below the 'Make top level'.

**Comment by MieWinstrup** [comment]
It seems like Travis failed because of missing test. 
Do I need to write a test for this feature? 
The other functions in the context-menu does not have a test and are excluded in the test (https://github.com/qgis/QGIS/blob/3e86804d644f453a9ce449d4684e6d819e685793/tests/code_layout/acceptable_missing_doc.py)

**Comment by nyalldawson** [comment]
@miewinstrup

A couple of years back it was decided that ALL new methods added to API must have documentation strings, and there's a unit test to ensure this. That's what's failing here.

Because at that time we had some 4000+ undocumented members there is a "whitelist" of methods which get exempt from this test. That's why there's existing undocumented members in this class. But all new members MUST have doxygen documentation describing what they do. 

Members added to an existing class also require a 

    \since QGIS 3.2 

at the end of their docstrings so that plugin authors know which methods are safe to use.

**Comment by MieWinstrup** [comment]
@nyalldawson - thanks for your comments. I will add it to the PR.

**Comment by MieWinstrup** [comment]
I have now added the documentation strings and the code has passed all existing unit tests.
Do I need to do something else? Or just wait for someone to review my code?

**Comment by nyalldawson** [comment]
Looking good! You'll need to fix the merge conflicts after #6665 was merged.

Could you add a unit test for this too?

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp [hunk 1]
```diff
@@ -342,6 +349,27 @@ void QgsLayerTreeViewDefaultActions::makeTopLevel()
   }
 }
 
+void QgsLayerTreeViewDefaultActions::moveToTop()
+{
+  QMap <QgsLayerTreeGroup *, int> groupInsertIdx;
+  int insertIdx;
+  Q_FOREACH ( QgsLayerTreeNode *n, mView->selectedNodes() )
```
Comment (elpaso) [inline src/gui/layertree/qgslayertreeviewdefaultactions.cpp]: We are trying to not use `Q_FOREACH` in new code, would you please use a `for` range loop here?

Comment (Gustry) [reply src/gui/layertree/qgslayertreeviewdefaultactions.cpp]: @MieWinstrup FYI: https://github.com/qgis/QGIS-Enhancement-Proposals/issues/103

**Comment by MieWinstrup** [comment]
@nyalldawson - thanks for your comment! 
Regarding creating a unittest for this feature - I haven't been able to find a unittest that interacts with QgsLayerTreeView and how to connect it with QgsLayerTreeViewDefaultActions, so I am not sure how best to write the unittest. Is there a way to get the current LayerTreeView when writing the test? Or should the test be written in the python layer a bit like QGIS/tests/src/python/test_qgslayertree.py?

**Comment by nyalldawson** [comment]
@MieWinstrup 

#6729 implements some basic unit tests for QgsLayerTreeView/QgsLayerTreeViewDefaultActions . If that passes I'll merge and you can use it as a basis for the new tests

**Comment by slarosa** [comment]
Yeah @nyalldawson, it is what I am trying to do here :-) to test move to topo action: https://github.com/slarosa/QGIS/commit/2de7f9c8d69274eb408101f7cb604c80e9086a2c

**Comment by nyalldawson** [comment]
@slarosa looks good! I'd just move it to the new layertreeview test file after #6729 is merged.

**Comment by MieWinstrup** [comment]
@slarosa and @nyalldawson  - thanks a lot for your input on test of the move-to-top-feature. Based on this, I have created a test that tests moving layers to the top. 

Travis fails on test regarding processing, which is unrelated to this pull request. Should I do anything about it?

File: tests/src/python/test_qgslayertreeview.py [hunk 2]
```diff
@@ -90,6 +90,16 @@ def testDefaultActions(self):
...
+    def testMoveToTopAction(self):
+        """Test move to top action"""
+        view = QgsLayerTreeView()
+        view.setModel(self.model)
+        actions = QgsLayerTreeViewDefaultActions(view)
+        self.assertEqual(self.project.layerTreeRoot().layerOrder(), [self.layer, self.layer2, self.layer3])
+        view.setCurrentLayer(self.layer3)
+        movetotop = actions.actionMoveToTop()
+        movetotop.trigger()
+        self.assertEqual(self.project.layerTreeRoot().layerOrder(), [self.layer3, self.layer, self.layer2])
```
Comment (slarosa) [inline tests/src/python/test_qgslayertreeview.py]: fix indentation (run ```scripts/prepare-commit.sh```)

Comment (MieWinstrup) [reply tests/src/python/test_qgslayertreeview.py]: I ran prepare-commit before my last commit and it said "nothing was modified". I think the indentation is the same as the test above. Can you be more specific on what needs to be changed?

Comment (slarosa) [reply tests/src/python/test_qgslayertreeview.py]: it is necessary an empty line and is what travis failure (code_layout build) was pointing:
```
--- tests/src/python/test_qgslayertreeview.py.prepare	2018-04-03 01:01:09.673477272 +0600
+++ tests/src/python/test_qgslayertreeview.py	2018-04-03 01:01:09.949480992 +0600
@@ -101,5 +101,6 @@
         movetotop.trigger()
         self.assertEqual(self.project.layerTreeRoot().layerOrder(), [self.layer3, self.layer, self.layer2])
 
+
 if __name__ == '__main__':
     unittest.main()
```

Comment (MieWinstrup) [reply tests/src/python/test_qgslayertreeview.py]: Thanks! New line added and I have pushed it to the old commit.

**Comment by nyalldawson** [comment]
Looking good! Could you also add a test for the group based logic? That's the trickier logic, and the part which is more likely to break if refactoring happens in future.

**Comment by MieWinstrup** [comment]
@nyalldawson - yes I will give it a try. As I see it, I can't use layerorder() for determining the order of groups and layers. Therefore I think I will need to run through children() to get the order of treenodes - will that be okay in the test?

**Comment by MieWinstrup** [comment]
@nyalldawson - tests for groups are now added :)

**Comment by nyalldawson** [comment]
Perfect! If Travis is happy I'll merge

**Pull request merged**